tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -1724,13 +1724,39 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
                                 PatternRewriter& rewriter) const override {
     auto type = op.getType().dyn_cast<RankedTensorType>();
     auto operandType = op.operand().getType().dyn_cast<RankedTensorType>();
-    if (!type || !type.hasStaticShape() || !operandType ||
+    auto outputDimOp = op.output_dimensions().getDefiningOp();
+    auto outputDimType =
+        op.output_dimensions().getType().dyn_cast<RankedTensorType>();
+    if (!type || !operandType || !outputDimType ||
         !operandType.hasStaticShape()) {
-      return rewriter.notifyMatchFailure(op, "requires static shape");
+      return rewriter.notifyMatchFailure(op, "requires operand static shape");
     }
-    rewriter.replaceOpWithNewOp<BroadcastInDimOp>(
-        op, op.getType(), op.operand(), op.broadcast_dimensions());
-    return success();
+    if (type.hasStaticShape()) {
+      rewriter.replaceOpWithNewOp<BroadcastInDimOp>(op, type, op.operand(),
+                                                    op.broadcast_dimensions());
+      return success();
+    } else if (outputDimOp &&
+               outputDimOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+      SmallVector<int64_t, 4> outputShape;
+      DenseIntElementsAttr shapeAttr;
+      if (auto constOp = llvm::dyn_cast<mhlo::ConstOp>(outputDimOp)) {
+        shapeAttr = constOp.value().cast<DenseIntElementsAttr>();
+      } else if (auto shapeOp =
+                     llvm::dyn_cast<shape::ConstShapeOp>(outputDimOp)) {
+        shapeAttr = shapeOp.getShapeAttr();
+      } else {
+        return rewriter.notifyMatchFailure(op, "unsupport constant");
+      }
+      for (APInt shape : shapeAttr) {
```

Comment:
I would think we could write `for (int64_t shape : shapeAttr.getValues<int64_t>())` ?

Reply:
Maybe we couldn't write like this. There is `%0 = mhlo.constant dense<[4, 32]> : tensor<2xi32>`. 
`getValues<int64_t>` will assert failing in this pattern.

Reply:
Well `getValues<int32_t>()` then :)
(I didn't check the actual element type used before)

Reply:
Also, there has pattern: `%0 = shape.const_shape [4, 32] : tensor<2xindex>`.  
So I think using `APInt` is suitable.